contracts/src/protocol/OmniPortal.sol

```diff
@@ -145,11 +145,54 @@ contract OmniPortal is
         require(msg.sender != address(this), "OmniPortal: portal cannot xcall");
         require(msg.value >= feeFor(destChainId, data, gasLimit), "OmniPortal: insufficient fee");
 
+        (uint32 _xcallOffset, uint64 _xblockOffset, uint128 _sourceBlockNumber) = _parseLastXCallID();
+
+        if (uint128(block.number) > _sourceBlockNumber) {
+            _sourceBlockNumber = uint128(block.number);
+            _xblockOffset = uint64(_xblockOffset + 1);
+            _xcallOffset = 0;
+        } else {
+            _xcallOffset += 1;
+            _xblockOffset = _xblockOffset;
+            _sourceBlockNumber = uint128(block.number);
```

Comment:
Necessary to keep gas usage between `else` branch equal to `if` branch.

This is messy I don't love it